File: smolml/models/transformer.py
"""Transformer baseline: pre-norm, RMSNorm, RoPE, causal self-attention.

Plain and honestly counted — the bar every candidate must clear. Sizes are
configurable so the same architecture spans tiny smoke runs to larger budgets.
The FFN is a 2-layer GELU MLP (``d -> d_ff -> d``), which is exactly what
:meth:`Transformer.flops` charges.
"""

from dataclasses import dataclass

import torch
import torch.nn.functional as F
from torch import nn

from smolml.data.corpus import VOCAB_SIZE
from smolml.flops import MAC_FLOPS, FlopBreakdown, causal_attention_flops, linear_flops
from smolml.models.registry import DecodeState, LanguageModel, register_model


@dataclass
class TransformerConfig:
    """Hyperparameters for the transformer baseline.

    ``d_ff`` defaults to ``4 * d_model``; ``d_model`` must be divisible by
    ``n_heads`` and the resulting head dim must be even (RoPE rotates pairs).
    """

    d_model: int = 128
    n_layers: int = 4
    n_heads: int = 4
    d_ff: int | None = None
    max_seq_len: int = 256
    vocab_size: int = VOCAB_SIZE
    rope_base: float = 10000.0
    dropout: float = 0.0
    tie_embeddings: bool = True

    def __post_init__(self):
        if self.d_ff is None:
            self.d_ff = 4 * self.d_model
        if self.d_model % self.n_heads != 0:
            raise ValueError(f"d_model {self.d_model} not divisible by n_heads {self.n_heads}")
        if (self.d_model // self.n_heads) % 2 != 0:
            raise ValueError("head dim (d_model / n_heads) must be even for RoPE")


class RMSNorm(nn.Module):
    """Root-mean-square layer norm (no mean subtraction, no bias)."""

    def __init__(self, dim: int, eps: float = 1e-6):
        super().__init__()
        self.eps = eps
        self.weight = nn.Parameter(torch.ones(dim))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        norm = torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + self.eps)
        return x * norm * self.weight


def build_rope_cache(seq_len: int, head_dim: int, base: float) -> tuple[torch.Tensor, torch.Tensor]:
    """Precompute (cos, sin) of shape ``(seq_len, head_dim)`` for rotary embeddings."""
    inv_freq = 1.0 / (base ** (torch.arange(0, head_dim, 2, dtype=torch.float32) / head_dim))
    t = torch.arange(seq_len, dtype=torch.float32)
    freqs = torch.outer(t, inv_freq)
    emb = torch.cat((freqs, freqs), dim=-1)
    return emb.cos(), emb.sin()


def _rotate_half(x: torch.Tensor) -> torch.Tensor:
    x1, x2 = x.chunk(2, dim=-1)
    return torch.cat((-x2, x1), dim=-1)


def apply_rope(x: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor) -> torch.Tensor:
    """Apply rotary position embedding to ``x`` of shape ``(B, n_heads, T, head_dim)``."""
    cos = cos[None, None, :, :]
    sin = sin[None, None, :, :]
    return x * cos + _rotate_half(x) * sin


class CausalSelfAttention(nn.Module):
    """Multi-head causal self-attention with RoPE on queries and keys."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()
        self.n_heads = cfg.n_heads
        self.head_dim = cfg.d_model // cfg.n_heads
        self.dropout = cfg.dropout
        self.qkv = nn.Linear(cfg.d_model, 3 * cfg.d_model, bias=False)
        self.proj = nn.Linear(cfg.d_model, cfg.d_model, bias=False)

    def forward(self, x: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor) -> torch.Tensor:
        b, t, c = x.shape
        q, k, v = self.qkv(x).split(c, dim=2)
        q = q.view(b, t, self.n_heads, self.head_dim).transpose(1, 2)
        k = k.view(b, t, self.n_heads, self.head_dim).transpose(1, 2)
        v = v.view(b, t, self.n_heads, self.head_dim).transpose(1, 2)
        q = apply_rope(q, cos, sin)
        k = apply_rope(k, cos, sin)
        out = F.scaled_dot_product_attention(
            q, k, v, is_causal=True, dropout_p=self.dropout if self.training else 0.0
        )
        out = out.transpose(1, 2).reshape(b, t, c)
        return self.proj(out)

    def decode_step(
        self,
        x: torch.Tensor,
        cos: torch.Tensor,
        sin: torch.Tensor,
        kv: tuple[torch.Tensor, torch.Tensor] | None,
    ) -> tuple[torch.Tensor, tuple[torch.Tensor, torch.Tensor]]:
        """Single-token incremental step with a KV cache.

        ``x`` is ``(B, 1, C)`` (the one new token); ``cos``/``sin`` are the rotary
        factors at its absolute position; ``kv`` is the cached (keys, values) from
        earlier positions, or ``None``. The new query attends to all cached keys
        plus itself (no mask needed — every cached key is in the past), exactly
        reproducing the last row of the full causal attention. Returns the
        projected output and the grown cache.
        """
        b, _, c = x.shape
        q, k, v = self.qkv(x).split(c, dim=2)
        q = q.view(b, 1, self.n_heads, self.head_dim).transpose(1, 2)
        k = k.view(b, 1, self.n_heads, self.head_dim).transpose(1, 2)
        v = v.view(b, 1, self.n_heads, self.head_dim).transpose(1, 2)
        q = apply_rope(q, cos, sin)
        k = apply_rope(k, cos, sin)
        if kv is not None:
            k = torch.cat([kv[0], k], dim=2)
            v = torch.cat([kv[1], v], dim=2)
        out = F.scaled_dot_product_attention(q, k, v, is_causal=False)
        out = out.transpose(1, 2).reshape(b, 1, c)
        return self.proj(out), (k, v)


class MLP(nn.Module):
    """Position-wise feed-forward: Linear -> GELU -> Linear."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()
        self.fc1 = nn.Linear(cfg.d_model, cfg.d_ff, bias=False)
        self.fc2 = nn.Linear(cfg.d_ff, cfg.d_model, bias=False)
        self.drop = nn.Dropout(cfg.dropout)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.drop(self.fc2(F.gelu(self.fc1(x))))


class Block(nn.Module):
    """Pre-norm transformer block: x + attn(norm(x)); x + mlp(norm(x))."""

    def __init__(self, cfg: TransformerConfig):
        super().__init__()
        self.norm1 = RMSNorm(cfg.d_model)
        self.attn = CausalSelfAttention(cfg)
        self.norm2 = RMSNorm(cfg.d_model)
        self.mlp = MLP(cfg)

    def forward(self, x: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor) -> torch.Tensor:
        x = x + self.attn(self.norm1(x), cos, sin)
        x = x + self.mlp(self.norm2(x))
        return x

    def decode_step(
        self,
        x: torch.Tensor,
        cos: torch.Tensor,
        sin: torch.Tensor,
        kv: tuple[torch.Tensor, torch.Tensor] | None,
    ) -> tuple[torch.Tensor, tuple[torch.Tensor, torch.Tensor]]:
        a, new_kv = self.attn.decode_step(self.norm1(x), cos, sin, kv)
        x = x + a
        x = x + self.mlp(self.norm2(x))
        return x, new_kv


@register_model("transformer")
class Transformer(LanguageModel):
    """The byte-level transformer baseline."""

    def __init__(self, config: TransformerConfig):
        super().__init__()
        self.config = config
        self.tok_emb = nn.Embedding(config.vocab_size, config.d_model)
        self.blocks = nn.ModuleList([Block(config) for _ in range(config.n_layers)])
        self.norm_f = RMSNorm(config.d_model)
        self.head = nn.Linear(config.d_model, config.vocab_size, bias=False)
        if config.tie_embeddings:
            self.head.weight = self.tok_emb.weight
        head_dim = config.d_model // config.n_heads
        cos, sin = build_rope_cache(config.max_seq_len, head_dim, config.rope_base)
        self.register_buffer("rope_cos", cos, persistent=False)
        self.register_buffer("rope_sin", sin, persistent=False)
        self.apply(self._init_weights)

    @staticmethod
    def _init_weights(module: nn.Module):
        if isinstance(module, nn.Linear):
            nn.init.normal_(module.weight, mean=0.0, std=0.02)
            if module.bias is not None:
                nn.init.zeros_(module.bias)
        elif isinstance(module, nn.Embedding):
            nn.init.normal_(module.weight, mean=0.0, std=0.02)

    def forward(self, idx: torch.Tensor) -> torch.Tensor:
        _, t = idx.shape
        if t > self.config.max_seq_len:
            raise ValueError(f"seq_len {t} exceeds max_seq_len {self.config.max_seq_len}")
        x = self.tok_emb(idx)
        cos, sin = self.rope_cos[:t], self.rope_sin[:t]
        for block in self.blocks:
            x = block(x, cos, sin)
        x = self.norm_f(x)
        return self.head(x)

    def flops(self, seq_len: int) -> FlopBreakdown:
        cfg = self.config
        t, d = seq_len, cfg.d_model
        per_layer = (
            linear_flops(t, d, 3 * d)  # qkv projection
            + linear_flops(t, d, d)  # output projection
            + linear_flops(t, d, cfg.d_ff)  # ffn up
            + linear_flops(t, cfg.d_ff, d)  # ffn down
            + causal_attention_flops(t, d)  # scores + value mixing
        )
        forward = cfg.n_layers * per_layer + linear_flops(t, d, cfg.vocab_size)
        return FlopBreakdown.from_forward(forward)

    # --- Prequential decode (KV-cache override; O(context_len) per byte) -------

    def init_prequential_state(self) -> DecodeState:
        # First byte predicted from a uniform prior; one empty KV slot per layer.
        return DecodeState(
            next_logits=torch.zeros(self.config.vocab_size),
            cache=[None] * self.config.n_layers,
        )

    def observe(
        self, state: DecodeState, token: int, pos: int
    ) -> tuple[DecodeState, FlopBreakdown]:
        cfg = self.config
        if pos >= cfg.max_seq_len:
            raise ValueError(
                f"prequential position {pos} reaches max_seq_len {cfg.max_seq_len}; "
                "size max_seq_len >= eval-stream length (sliding-window decode is future work)"
            )
        device = self.rope_cos.device
        x = self.tok_emb(torch.tensor([[token]], dtype=torch.long, device=device))
        cos, sin = self.rope_cos[pos : pos + 1], self.rope_sin[pos : pos + 1]
        new_cache: list[tuple[torch.Tensor, torch.Tensor]] = []
        with torch.no_grad():
            for block, kv in zip(self.blocks, state.cache, strict=True):
                x, nkv = block.decode_step(x, cos, sin, kv)
                new_cache.append(nkv)
            logits = self.head(self.norm_f(x))[0, -1].detach()
        # This step attended pos+1 keys (the cache of pos plus the new token).
        return DecodeState(next_logits=logits, cache=new_cache), self.decode_step_flops(pos + 1)

    def decode_step_flops(self, context_len: int) -> FlopBreakdown:
        """Incremental per-byte prediction cost with a KV cache: O(d^2) projections
        for the one new token + O(context_len * d) attention (1 query vs
        ``context_len`` cached keys). Forward-only (inference)."""
        cfg = self.config
        d = cfg.d_model
        per_layer = (
            linear_flops(1, d, 3 * d)  # qkv for the single new token
            + linear_flops(1, d, d)  # output projection
            + linear_flops(1, d, cfg.d_ff)  # ffn up
            + linear_flops(1, cfg.d_ff, d)  # ffn down
            + 2 * MAC_FLOPS * d * context_len  # Q·Kᵀ + softmax·V: 1 query × context_len keys
        )
        forward = cfg.n_layers * per_layer + linear_flops(1, d, cfg.vocab_size)
        return FlopBreakdown(forward=forward, backward=0)

    @classmethod
    def from_config(cls, config: dict[str, object]) -> "Transformer":
        return cls(TransformerConfig(**config))

File: docs/learning/public/js/models/transformer.js
/* transformer — byte-level pre-norm RMSNorm/RoPE causal transformer, faithful
 * port of smolml/models/transformer.py decode for the prequential step.
 *
 * The Python step runs an incremental KV-cache decode while the context is below
 * max_seq_len (the GROWING regime) and a full windowed recompute once it is full
 * (the SLIDING regime, which drops the cache and resets RoPE positions to
 * 0..win-1). Both regimes equal a full causal forward over the last
 * min(len, max_seq_len) bytes read at the final position — so this port simply
 * does that forward every step and matches Python in BOTH regimes. Weights are
 * loaded from the tiny trained export (float32 base64); compute is float64.
 *
 * API: create({ config, weights }); step(state, byte, pos) -> { probs, state }.
 *   weights = transformer.weights.json (see export_demo_fixtures.py). */
(function (global) {
  "use strict";
  var H = global.SmolDemos;
  var MODELS = (global.SmolModels = global.SmolModels || {});
  var EPS = 1e-6;

  // y[t,o] = Σ_i x[t,i]·W[o,i]   (torch nn.Linear: y = x · Wᵀ, W is (out,in)).
  function linear(x, T, din, W, dout, out) {
    for (var t = 0; t < T; t++) {
      var xb = t * din, ob = t * dout;
      for (var o = 0; o < dout; o++) {
        var wb = o * din, acc = 0.0;
        for (var i = 0; i < din; i++) acc += x[xb + i] * W[wb + i];
        out[ob + o] = acc;
      }
    }
    return out;
  }

  function rmsnorm(x, T, d, weight, out) {
    for (var t = 0; t < T; t++) {
      var b = t * d, ss = 0.0, j;
      for (j = 0; j < d; j++) { var v = x[b + j]; ss += v * v; }
      var nrm = 1.0 / Math.sqrt(ss / d + EPS);
      for (j = 0; j < d; j++) out[b + j] = x[b + j] * nrm * weight[j];
    }
    return out;
  }

  function create(opts) {
    opts = opts || {};
    var w = opts.weights;
    if (!w) throw new Error("transformer.create needs opts.weights (transformer.weights.json)");
    var cfg = w.config;
    var d = cfg.d_model, nh = cfg.n_heads, hd = d / nh, dff = cfg.d_ff;
    // RoPE (cos,sin) for positions 0..max_seq_len-1 — recomputed (not exported).
    var P = cfg.max_seq_len, half = hd / 2;
    var invFreq = new Float64Array(half);
    for (var i = 0; i < half; i++) invFreq[i] = 1.0 / Math.pow(cfg.rope_base, (2 * i) / hd);
    var cos = new Float64Array(P * hd), sin = new Float64Array(P * hd);
    for (var t = 0; t < P; t++) {
      for (var k = 0; k < half; k++) {
        var ang = t * invFreq[k];
        var cv = Math.cos(ang), sv = Math.sin(ang);
        cos[t * hd + k] = cv; cos[t * hd + half + k] = cv;
        sin[t * hd + k] = sv; sin[t * hd + half + k] = sv;
      }
    }
    var blocks = w.blocks.map(function (bl) {
      return {
        norm1: H.decodeF32(bl.norm1), qkv: H.decodeF32(bl.qkv), proj: H.decodeF32(bl.proj),
        norm2: H.decodeF32(bl.norm2), fc1: H.decodeF32(bl.fc1), fc2: H.decodeF32(bl.fc2),
      };
    });
    return {
      cfg: cfg, d: d, nh: nh, hd: hd, half: half, dff: dff, V: cfg.vocab_size,
      emb: H.decodeF32(w.tok_emb), normF: H.decodeF32(w.norm_f), blocks: blocks,
      cos: cos, sin: sin,
      window: [], // rolling byte window, capped at max_seq_len
    };
  }

  function applyRope(vec, base, cos, sin, cb, half, hd) {
    // in place: result[m<half] = v[m]·cos[m] - v[m+half]·sin[m];
    //           result[m>=half] = v[m]·cos[m] + v[m-half]·sin[m].
    for (var m = 0; m < half; m++) {
      var a = vec[base + m], b = vec[base + half + m];
      var cl = cos[cb + m], sl = sin[cb + m];
      vec[base + m] = a * cl - b * sl;
      vec[base + half + m] = b * cos[cb + half + m] + a * sin[cb + half + m];
    }
  }

  function forwardLast(state, tokens) {
    var T = tokens.length, d = state.d, nh = state.nh, hd = state.hd, half = state.half;
    var dff = state.dff, V = state.V, cos = state.cos, sin = state.sin;
    var x = new Float64Array(T * d), t, j;
    for (t = 0; t < T; t++) {
      var er = tokens[t] * d, xb = t * d;
      for (j = 0; j < d; j++) x[xb + j] = state.emb[er + j];
    }
    var nrm = new Float64Array(T * d), qkv = new Float64Array(T * 3 * d);
    var attn = new Float64Array(T * d), proj = new Float64Array(T * d);
    var h2 = new Float64Array(T * d), ff = new Float64Array(T * dff), mlp = new Float64Array(T * d);
    var scale = 1.0 / Math.sqrt(hd), scores = new Float64Array(T);

    for (var L = 0; L < state.blocks.length; L++) {
      var bl = state.blocks[L];
      rmsnorm(x, T, d, bl.norm1, nrm);
      linear(nrm, T, d, bl.qkv, 3 * d, qkv);
      // RoPE on q (cols 0..d) and k (cols d..2d) per head.
      for (t = 0; t < T; t++) {
        var cb = t * hd; // window positions are 0..T-1
        for (var hh = 0; hh < nh; hh++) {
          applyRope(qkv, t * 3 * d + hh * hd, cos, sin, cb, half, hd);
          applyRope(qkv, t * 3 * d + d + hh * hd, cos, sin, cb, half, hd);
        }
      }
      // Causal attention per head -> attn (T,d).
      for (var head = 0; head < nh; head++) {
        var off = head * hd;
        for (var qi = 0; qi < T; qi++) {
          var qbase = qi * 3 * d + off, mx = -Infinity, jj, mm;
          for (jj = 0; jj <= qi; jj++) {
            var kbase = jj * 3 * d + d + off, dot = 0.0;
            for (mm = 0; mm < hd; mm++) dot += qkv[qbase + mm] * qkv[kbase + mm];
            dot *= scale; scores[jj] = dot; if (dot > mx) mx = dot;
          }
          var ssum = 0.0;
          for (jj = 0; jj <= qi; jj++) { var e = Math.exp(scores[jj] - mx); scores[jj] = e; ssum += e; }
          var obase = qi * d + off;
          for (mm = 0; mm < hd; mm++) {
            var acc = 0.0;
            for (jj = 0; jj <= qi; jj++) acc += scores[jj] * qkv[jj * 3 * d + 2 * d + off + mm];
            attn[obase + mm] = acc / ssum;
          }
        }
      }
      linear(attn, T, d, bl.proj, d, proj);
      for (j = 0; j < T * d; j++) x[j] += proj[j];
      rmsnorm(x, T, d, bl.norm2, h2);
      linear(h2, T, d, bl.fc1, dff, ff);
      for (j = 0; j < T * dff; j++) ff[j] = H.gelu(ff[j]);
      linear(ff, T, dff, bl.fc2, d, mlp);
      for (j = 0; j < T * d; j++) x[j] += mlp[j];
    }
    // Final norm + tied head on the LAST position only.
    var last = new Float64Array(d), lb = (T - 1) * d, ss = 0.0;
    for (j = 0; j < d; j++) { var vv = x[lb + j]; ss += vv * vv; }
    var fn = 1.0 / Math.sqrt(ss / d + EPS);
    for (j = 0; j < d; j++) last[j] = x[lb + j] * fn * state.normF[j];
    var logits = new Float64Array(V);
    for (var b = 0; b < V; b++) {
      var eb = b * d, acc2 = 0.0;
      for (j = 0; j < d; j++) acc2 += last[j] * state.emb[eb + j];
      logits[b] = acc2;
    }
    return logits;
  }

  function step(state, revealedByte, pos) {
    var cap = state.cfg.max_seq_len;
    var win = state.window.concat([revealedByte]);
    if (win.length > cap) win = win.slice(win.length - cap);
    state.window = win;
    var logits = forwardLast(state, win);
    return { probs: H.softmax(logits), state: state };
  }

  MODELS.transformer = { create: create, step: step };
})(typeof globalThis !== "undefined" ? globalThis : this);
